### PR TITLE
Developer menu on version click

### PR DESCRIPTION
**Pull request by hannesa2**
1. The first click nothing happen
2. after you need 4 more clicks
3. now the developer menu appears (currently only log option)
4. further clicks point to github.com commit like before

You could use this switch for some other use cases too, but this is your decision (and your or additional PR)
Btw, I use it [here](https://github.com/hannesa2/owncloud-android/releases) to investigate why my [upload doesn't work](https://github.com/owncloud/android/issues/2394) and [AndroidX](https://github.com/owncloud/android/pull/2402) should crash

developermenu

### CODE REVIEW

**Comment by michaelstingl** [comment]
Awesome, I ❤️ it! 👍

BTW I granted you permissions to open branches in this repo.

**Comment by hannesa2** [comment]
I pointed out, it makes no sense, to show in `release` the log, when switch `isBeta()` prevents from logging.
So I merge new `isDeveloper()` (controlled by user) and `isBeta()`(controlled by build) with a second commit

**Comment by davigonz** [comment]
Thanks for this @hannesa2 , I guess we can now include logs for customers and PlayStore versions

**Comment by davigonz** [comment]
@hannesa2 I've just cloned your code and I'm able to see the Logs option (in debug and release) even before enabling the developer menu, any clue?

**Comment by hannesa2** [comment]
Fresh install?

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java [hunk 1]
```diff
@@ -266,18 +234,25 @@ public static String getUserAgent() {
...
-            }
-        } catch (PackageManager.NameNotFoundException e) {
-            e.printStackTrace();
+    public static boolean isDeveloper() {
+        return mDeveloper;
+    }
+
+    public void readIsDeveloper() {
+        mDeveloper = BuildConfig.DEBUG || PreferenceManager.getDefaultSharedPreferences(getApplicationContext())
+                .getInt(CLICK_DEV_MENU, 0) > CLICKS_NEEDED_TO_BE_DEVELOPER;
```
Comment (hannesa2) [inline owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: here is the magic

**Comment by hannesa2** [comment]
With commit **use isDevloper() everywhere instead of isBeta()** you see it in debug always 
https://github.com/owncloud/android/pull/2439/files#r259567889
and `release` can't be build locally with current repo.

It seems you have set the `System.env.OC_RELEASE_KEYSTORE` , I'm not.
Where is the release build ? I've the fear you build it local on your machine

**Comment by davigonz** [comment]
> With commit **use isDevloper() everywhere instead of isBeta()** you see it in debug always
https://github.com/owncloud/android/pull/2439/files#r259567889
and `release` can't be build locally with current repo.

I will try it again

**Comment by davigonz** [comment]
> `release` can't be build locally with current repo.

> It seems you have set the `System.env.OC_RELEASE_KEYSTORE` , I'm not.
Where is the release build ? I've the fear you build it local on your machine

You can build it by using release variant but you won't be allowed to install the generated apk without signing it. I signed the apk and installed it and as I said before, Logs option is visible in debug release even before enabling the developer menu.

You could give it a try after signing the apk with a test keystore.

**Comment by hannesa2** [comment]
Arghhh , it was working, please see animation in first post. 
Anyway, this repeating rebasing destroyed it, and I fixed it again. 
Please have a look

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java [hunk 2]
```diff
@@ -264,18 +233,25 @@ public static String getUserAgent() {
...
+
+        if (isDeveloper()) {
+
+            String dataFolder = getDataFolder();
+
+            // Set folder for store logs
+            Log_OC.setLogDataFolder(dataFolder);
+
+            Log_OC.startLogging(Environment.getExternalStorageDirectory().getAbsolutePath());
+            Log_OC.d(BuildConfig.BUILD_TYPE, "start logging " + BuildConfig.VERSION_NAME + " " + BuildConfig.COMMIT_SHA1);
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: Line too long

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java [hunk 3]
```diff
@@ -264,18 +233,25 @@ public static String getUserAgent() {
...
-            }
-        } catch (PackageManager.NameNotFoundException e) {
-            e.printStackTrace();
+    public static boolean isDeveloper() {
+        return mDeveloper;
+    }
+
+    public void readIsDeveloper() {
+        mDeveloper = BuildConfig.DEBUG || PreferenceManager.getDefaultSharedPreferences(getApplicationContext())
+                .getInt(CLICK_DEV_MENU, 0) > CLICKS_NEEDED_TO_BE_DEVELOPER;
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: Use a constant for the default clicks value = 0

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java [hunk 4]
```diff
@@ -264,18 +233,25 @@ public static String getUserAgent() {
...
+
+
+        if (isDeveloper()) {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: I would not include the code to start logging in here since this method (readIsDeveloper) is not expected to do that. Instead, return a boolean and call a different method to start logging from onCreate().

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java [hunk 5]
```diff
@@ -264,18 +233,25 @@ public static String getUserAgent() {
...
-                isBeta = true;
-            }
-        } catch (PackageManager.NameNotFoundException e) {
-            e.printStackTrace();
+    public static boolean isDeveloper() {
+        return mDeveloper;
+    }
+
+    public void readIsDeveloper() {
+        mDeveloper = BuildConfig.DEBUG || PreferenceManager.getDefaultSharedPreferences(getApplicationContext())
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: I miss here the beta case, beta users should be able to see the Logs option without enabling developer one in my opinion. I see the point of hiding the Logs option for normal users in Google Play, but not for beta ones.

File: owncloudApp/src/main/java/com/owncloud/android/authentication/AuthenticatorActivity.java [hunk 6]
```diff
@@ -257,7 +257,7 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
         /// protection against screen recording
-        if (!BuildConfig.DEBUG) {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/authentication/AuthenticatorActivity.java]: Please remove `import com.owncloud.android.BuildConfig;`

File: owncloudApp/src/main/java/com/owncloud/android/authentication/FingerprintManager.java [hunk 7]
```diff
@@ -75,7 +75,7 @@ protected FingerprintManager() {
     }
 
     public void onActivityCreated(Activity activity) {
-        if (!BuildConfig.DEBUG) {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/authentication/FingerprintManager.java]: Same

File: owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java [hunk 8]
```diff
@@ -57,14 +57,14 @@ public static PassCodeManager getPassCodeManager() {
...
 
-    private Long mTimestamp = 0l;
+    private Long mTimestamp = 0L;
     private int mVisibleActivitiesCounter = 0;
 
     protected PassCodeManager() {
     }
 
     public void onActivityCreated(Activity activity) {
-        if (!BuildConfig.DEBUG) {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java]: Same and in the rest of the classes where you have removed it

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java [hunk 9]
```diff
@@ -283,8 +288,7 @@ public void onCreate(Bundle savedInstanceState) {
             mPattern.setOnPreferenceChangeListener((preference, newValue) -> {
                 Intent intent = new Intent(getApplicationContext(), PatternLockActivity.class);
                 Boolean state = (Boolean) newValue;
-                SharedPreferences appPrefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java]: Good catch

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java [hunk 10]
```diff
@@ -468,21 +468,15 @@ public void onCreate(Bundle savedInstanceState) {
...
-        boolean loggerEnabled = getResources().getBoolean(R.bool.logger_enabled) ||
-                BuildConfig.DEBUG || MainApp.isBeta();
-        Preference pLogger = findPreference("logger");
-        if (pLogger != null) {
-            if (loggerEnabled) {
-                pLogger.setOnPreferenceClickListener(preference -> {
-                    Intent loggerIntent = new Intent(getApplicationContext(), LogHistoryActivity.class);
-                    startActivity(loggerIntent);
+        // show item(s) only when you are developer
+        mLogger = findPreference("logger");
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java]: I know the rest of the preferences are doing the same but "logger" should be in a constant and not hardcoded

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java [hunk 11]
```diff
@@ -513,15 +507,36 @@ public void onCreate(Bundle savedInstanceState) {
...
+                }
+                mAppPrefs.edit().putInt(MainApp.CLICK_DEV_MENU, clickCount + 1).apply();
+                ((MainApp) getApplication()).readIsDeveloper(); // read value to global variable
                 return true;
             });
         }
     }
 
+    private void showDeveloperItems(PreferenceCategory preferenceCategory) {
+
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java]: Delete this line

**Comment by davigonz** [comment]
@hannesa2 some changes requested

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java [hunk 12]
```diff
@@ -111,23 +112,12 @@ public void onCreate() {
...
-            Log_OC.d(BuildConfig.BUILD_TYPE, "start logging " + BuildConfig.VERSION_NAME);
-        }
-
         // register global protection with pass code, pattern lock and fingerprint lock
         registerActivityLifecycleCallbacks(new ActivityLifecycleCallbacks() {
 
             @Override
             public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
-                Log_OC.d(activity.getClass().getSimpleName(), "onCreate(Bundle) starting");
+                Log_OC.v(activity.getClass().getSimpleName(), "onCreate(Bundle) starting");
```
[the remark this answers is not in the document]
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: Ok, move to more spam. Done

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 2 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: This is the reason, why I vote for 150: there is enough place !
image

But you reduce to 120

This seems to be better ? Tastes are different
image

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 3 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: Now we have `private static final int DEFAULT_0 = 0;`

but the line below makes me much more doubt

image

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 5 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: What's about this ? 
image

File: owncloudApp/src/main/java/com/owncloud/android/authentication/FingerprintManager.java (hunk 7 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/authentication/FingerprintManager.java]: see above

File: owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java (hunk 8 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java]: there is no need of this anymore
image

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java (hunk 10 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java]: feel free to do it

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java (hunk 11 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/ui/activity/Preferences.java]: When I would search code for that, I expect a lot of this. But maybe I'm wrong
Done

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 4 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: This is my tested way, I will not change this anymore and I don't support your suggestion.
If you want to have it change, I pushed branch to 'upstream' remote, so you can change it and rebase ....

**Comment by hannesa2** [comment]
This is my tested way, I will not change this anymore.
If you want to have it change, I pushed branch to 'upstream' remote, so you can change it and rebase .... what ever you want

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: `Log_OC.v` would also include this log in the log file so it would be "spam" anyway. What I mean is that .v doesn't fit in this case because Log.v is for Verbose logging and is used when you want to output almost everything. On the other hand, Log.d is used to know the flow of the activities and this is the case.

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 2 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: We'd rather 120 because of how it looks in GH as well, the next screenshot is with more than 120 columns:

Screenshot_2019-03-14_at_09_38_49

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 3 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: Maybe you do not need to call it `DEFAULT_0` but `CLICKS_DEFAULT` since you are comparing it with `CLICKS_NEEDED_TO_BE_DEVELOPER`

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.java (hunk 5 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.java]: That's DEBUG build and beta releases will not go that way because are releases themselves. When we release a beta version, `mDeveloper` might be enabled by default to directly show the logs.

File: owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java (hunk 8 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java]: That's what I mean, if is not needed anymore, please delete the `import com.owncloud.android.BuildConfig;` dependency at the top of the classes.

**Comment by davigonz** [comment]
> This is my tested way, I will not change this anymore.
If you want to have it change, I pushed branch to 'upstream' remote, so you can change it and rebase .... what ever you want

> I'm done here

Sorry @hannesa2 but this is not how contributions work, there's a process that you most likely already know and if any contributor wants to have his code merged, first it needs an approved code review and to achieve that, we normally request some changes. And being honest, all of our contributors do always agree with this.

Not all the changes requested are mandatory of course and are all of them are opened to discussion but leaving a PR half done and expecting that anyone of the core developers is going to finish your contribution is not the best way to proceed because we have spent time reviewing your code to make it better - or at least, that's the intention - when we could directly apply the changes without expending time in adding comments to the review.

We've already performed some changes in our flow to make the contributors feel more comfortable and work less, now you have permissions to create branches directly in the ownCloud Android repository and we were willing to rebase your contributions even knowing that is not the best approach - the owner of each PR should take care of rebases, including contributors, and solve conflicts if needed, is the process we follow from time ago -.

It's a fact that you are using your spare time to contribute here and we are very grateful but as developers it does not mean that we have to finish your contributions or you are allowed to skip some steps of our process.

So now let's resume the work in the code review:
https://github.com/owncloud/android/pull/2439#discussion_r265095616  and https://github.com/owncloud/android/pull/2439#discussion_r265099766 would be necessary - make the code cleaner and readable -  to merge this contribution. 
If you do not agree with https://github.com/owncloud/android/pull/2439#discussion_r265095616 for instance, please give us an argument cause you just wrote this.

Screenshot 2019-03-14 at 10 25 07

Thanks

**Comment by hannesa2** [comment]
No problem, currently I'm not in the mod to do all suggestions to get merged. 

Maybe this will change again, but currently you have my suggestions as PR. You can use it as source of an ideas, accept it, or decline it.

I learned, when I've serious issues (the broken sync) **I've to do it by myself**. Nobody does the paining deeper bug fixing. -that's a fact-

All my remaining PRs are just by-products of this sync-fix to help **me** (nobody else) to investigate what's wrong.

One big thing I've to fix https://github.com/owncloud/android/issues/2460 (help yourself seems the only way) and I guess the world should have this bugfix too. 

But when I get as a review issue, **to remove an empty line (!)** I feel the strong trigger to change my work mode

So I do **my** changes, show it to this repo, to let you participate. Maybe I should rethink this too, to avoid misunderstandings

But I've to save my time, that's the part of volunteers.

File: owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java (hunk 8 above)
Comment (hannesa2) [reply owncloudApp/src/main/java/com/owncloud/android/authentication/PassCodeManager.java]: Ok, this I missunderstood

**Comment by jesmrec** [comment]
Just to clarify:

Your time is your time and you plan it, of course.  When the oC team asks you for changes, this is not a matter of time. I mean, you can do it whenever you can/want. If today you are busy, do it the next week or the next month, this is not a problem. Community works in this way. 

Priorities are set by the team, taking in account many factors. We take care of the incoming PRs when the whole team agrees to do it. If any PR is not attended at any moment is because other stuff with more priority is being attended. 

Code reviews try to get the best code quality. Everyone has his own idea about "quality", i know. But things like "remove an empty line" is asked for everyone in the community and also internally in the team as part of the code review. Will not take exceptions. CR is mandatory to pass, and discussions about the suggestions are welcome, making not all the changes to be always done. But skipping or refusing... is not the solution.

The team (and specially @davigonz) is fully engaged with community as always did. Product MUST be the target for team and community,  and not "me", "myself" and similars.  All of us together, by respecting ourselves, will make our lives easier.

Hoping for a solution regarding the current PR.

**Comment by davigonz** [comment]
> I learned, when I've serious issues (the broken sync) I've to do it by myself. Nobody does the paining deeper bug fixing. -that's a fact-

> All my remaining PRs are just by-products of this sync-fix to help me (nobody else) to investigate what's wrong.

> One big thing I've to fix is #2460 (help yourself seems the only way) and I guess the world should have this bugfix too.

You just need to have a look at https://github.com/owncloud/android/issues/2394#issuecomment-451081692 and the rest of the comments I put on that issue to realise we were trying to help you, asking you for more information to find out the cause of your problem. 

In addition, we tried to reproduce the issue in our side multiple times (https://github.com/owncloud/android/issues/2394#issuecomment-450815173) and was not possible, that's why we did not start the "deeper bug fixing". Then your broken sync was fixed by something we do not know clearly yet, as you commented in https://github.com/owncloud/android/issues/2394#issuecomment-462174779. 

Another example of our help => https://github.com/owncloud/android/issues/2393#issuecomment-467953043

> But when I get as a review issue, to remove an empty line (!) #2439 (comment) I feel the strong trigger to change my work mode

You have picked the most insignificant change there, what about [this](https://github.com/owncloud/android/pull/2439/files/80bf9a87d4e0a83f3d605e8d01b2cf7b8c4024d6#diff-8d74822b24b5437f88e813b9c5488ed9) where we could improve the code readability or [this](https://github.com/owncloud/android/pull/2439#discussion_r265100323) to remove unneeded dependencies? That's the aim of CRs.

> So I do my changes, show it to this repo, to let you participate. Maybe I should rethink this too, to avoid misunderstandings

We have participated in many of you contributions, reviewed and merge [25 contributions](https://github.com/owncloud/android/pulls?q=is%3Apr+is%3Aclosed+author%3Ahannesa2) in the ownCloud Android app and [16](https://github.com/owncloud/android-library/pulls?q=is%3Apr+is%3Aclosed+author%3Ahannesa2) in the ownCloud Android library.

To sum up, we would like to keep including your contributions but first we need mutual respect as @jesmrec suggests and discrediting our work here or suggesting we do not help is the worst way to achieve that in my humble opinion.

**Comment by michaelstingl** [comment]
> Maybe this will change again, but currently you have my suggestions as PR. You can use it as source of an ideas, accept it, or decline it.

Thank you very much! We'll consider to finish this and bring it to the final app. Maybe 2.11.0 – let's see…

**Comment by davigonz** [comment]
Let's merge this via https://github.com/owncloud/android/pull/2502, thanks @hannesa2 for the contribution
